zerocopy-derive/src/enums.rs

```diff
@@ -0,0 +1,550 @@
+        Ident::new(&format!("{}{}", self.variant_prefix, variant), variant.span())
+    }
+}
+
+/// Generates an implementation of `is_bit_valid` for the given enum input.
+pub(crate) fn derive_is_bit_valid(
+    reprs: &[EnumRepr],
+    ident: &Ident,
+    generics: &Generics,
+    data: &DataEnum,
+) -> TokenStream {
+    let idents = RawIdents::for_enum(ident);
+
+    if reprs.contains(&EnumRepr::C) {
+        derive_is_bit_valid_c(data, generics, &idents)
+    } else {
+        derive_is_bit_valid_primitive(data, generics, discriminant_repr(reprs), &idents)
+    }
+}
+
+pub(crate) fn discriminant_repr(reprs: &[EnumRepr]) -> &EnumRepr {
+    // This returns the first non-C, non-align repr. We know that `reprs` will
+    // contain at least `C` if not also in integer because we check the valid
+    // reprs in the derive which calls this function. If we don't find a non-C,
+    // non-align repr then we know that it was `repr(C)` all along.
+    reprs.iter().find(|&r| !matches!(r, EnumRepr::C | EnumRepr::Align(_))).unwrap_or(&EnumRepr::C)
+}
+
+pub(crate) fn generate_discriminant_type(
+    data: &DataEnum,
+    repr: &EnumRepr,
+    name: &Ident,
+) -> TokenStream {
+    let variants = data.variants.iter().map(|v| {
+        let ident = &v.ident;
+        if let Some((eq, discriminant)) = &v.discriminant {
+            quote! { #ident #eq #discriminant }
+        } else {
+            quote! { #ident }
+        }
+    });
+
+    // This raw discriminant has the same repr and explicit discriminants as the
+    // original enum, but no fields.
+    quote! {
+        #[repr(#repr)]
+        #[allow(dead_code)]
+        enum #name {
+            #(#variants),*
+        }
+    }
+}
+
+fn generate_discriminant_consts(data: &DataEnum, idents: &RawIdents) -> TokenStream {
+    let discriminant = &idents.discriminant;
+
+    let consts = data.variants.iter().map(|v| {
+        let ident = &v.ident;
+        let const_ident = idents.discriminant_const(ident);
+
+        // NOTE: We're using `[u8; N]` as the type of the discriminant contants
+        // here because the sizes of `#[repr(C)]` fieldless enums is not well-
+        // specified. There _might_ be an opportunity to generate better code
+        // for enums with the primitive representation, but it also may not make
+        // any difference.
+        quote! {
+            // SAFETY: The raw discriminant enum is a C-style enum, and all
+            // C-style enums are represented as simple integers. All integers
+            // can be transmuted to their constituent bytes.
+            const #const_ident: [u8; core_reexport::mem::size_of::<#discriminant>()] = unsafe {
+                core_reexport::mem::transmute(#discriminant::#ident)
+            };
+        }
+    });
+
+    quote! {
+        #(#consts)*
+    }
+}
+
+fn generate_variant_struct(
+    variant: &Variant,
+    generics: &Generics,
+    idents: &RawIdents,
+    with_tag: bool,
```

Comment:
Is this is only invoked with `with_tag: false`?

Reply:
Copy-paste error, fixed now

Reply:
Follow-up: Can we change this to an enum so it's self-documenting at the call sites? E.g., `generate_variant_struct(..., TagLocation::Outer)` or something.

Reply:
Switching to outer tag / inner tag to unify repr(C) and repr(int) code paths make this obsolete